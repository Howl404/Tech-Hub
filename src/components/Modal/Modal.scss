.modal {
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.4);
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  opacity: 0;
  pointer-events: none;
  transition: all 0.5s;
  // flex-direction: column;
}

.modal.active__modal {
  opacity: 1;
  pointer-events: all;
}

.modal__content {
  padding: 20px;
  border-radius: 10px;
  background-color: white;
  // width: 400px;
  // height: 450px;
  transform: scale(0.5);
  transition: all 0.4s;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.modal__content.active__modal {
  transform: scale(1);
}

.btn__save {
  color: #fff;
  text-align: center;
  font-family: 'Oswald', sans-serif;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  width: 150px;
  height: 50px;
  background: black;
  text-transform: uppercase;
  &:hover {
    background: rgba(0, 0, 0, 0.685);
  }
  &:disabled {
    cursor: auto;
    background-color: rgba(0, 0, 0, 0.101);
  }
}
