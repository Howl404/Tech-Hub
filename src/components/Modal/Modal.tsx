/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useEffect } from 'react';
import './Modal.scss';
import { FaRegSave } from 'react-icons/fa';
import { PostalCodePattern } from '@src/interfaces/Register';
import { sendData } from '@src/services/AuthService/AuthService';
import FormInput from '../FormInput/FormInput';

interface ModalType {
  active: boolean;
  userId: string;
  setActive: (logic: boolean) => void;
  city: string;
  postalCode: string;
  country: string;
  streetName: string;
  addressId: string;
  selectedData: {
    city: string;
    postalCode: string;
    country: string;
    streetName: string;
  };
  setSelectedData: React.Dispatch<
    React.SetStateAction<{
      city: string;
      postalCode: string;
      country: string;
      streetName: string;
    }>
  >;
}

const postalCodePattern: PostalCodePattern = {
  US: '\\d{5}-\\d{4}|\\d{5}',
  RU: '\\d{6}',
  GB: '[A-Za-z]{1,2}\\d{1,2}[A-Za-z]?\\s?\\d[A-Za-z]{2}',
  DE: '\\d{5}',
  FR: '\\d{5}',
};

function Modal({
  active,
  setActive,
  city,
  userId,
  postalCode,
  country,
  streetName,
  selectedData,
  setSelectedData,
  addressId,
}: ModalType): JSX.Element {
  const handleCountryChange = (event: React.ChangeEvent<HTMLSelectElement>): void => {
    const { id, value } = event.target;
    setSelectedData({ ...selectedData, [id]: value });
  };

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { id, value } = event.target;
    setSelectedData({ ...selectedData, [id]: value });
  };

  return (
    <div className={active ? 'modal active__modal' : 'modal'} onClick={(): void => setActive(false)}>
      <div
        className={active ? 'modal__content active__modal' : 'modal__content'}
        onClick={(e): void => e.stopPropagation()}
      >
        <div className="form-input">
          <label htmlFor="country">
            Country
            <select id="country" name="country" value={country} onChange={handleCountryChange}>
              <option value="" label="Select a country ... ">
                Select a country ...
              </option>
              <option value="US" label="United States">
                United States
              </option>
              <option value="RU" label="Russia">
                Russia
              </option>
              <option value="GB" label="United Kingdom">
                United Kingdom
              </option>
              <option value="DE" label="Germany">
                Germany
              </option>
              <option value="FR" label="France">
                France
              </option>
            </select>
          </label>
        </div>
        <FormInput
          label="City"
          errorMessage="City is not valid"
          onChange={handleInputChange}
          id="city"
          type="text"
          pattern="[A-Za-z]+"
          title="Must contain only letters"
          value={city}
        />
        <FormInput
          label="Postal Code"
          errorMessage="Invalid Postal Code"
          onChange={handleInputChange}
          id="postalCode"
          type="text"
          pattern={postalCodePattern[country] || '.*'}
          title="Must be a valid postal code of a selected country"
          value={postalCode}
        />
        <FormInput
          label="Street name"
          errorMessage="Less than 1 character"
          onChange={handleInputChange}
          id="streetName"
          type="text"
          pattern=".*"
          title="Must contain more than 1 character"
          value={streetName}
        />
        <button
          type="button"
          className="btn__save"
          onClick={(): Promise<void> =>
            sendData(selectedData, userId, addressId).then((item) => {
              const date = item.addresses.filter((adress) => adress.id === addressId)[0];
              setSelectedData({
                city: date.city,
                postalCode: date.postalCode,
                country: date.country,
                streetName: date.streetName,
              });
              setActive(false);
            })
          }
        >
          save <FaRegSave />
        </button>
      </div>
    </div>
  );
}

export default Modal;
