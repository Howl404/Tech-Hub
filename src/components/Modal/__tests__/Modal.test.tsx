import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import Modal, { ModalType } from '../Modal';
import '@testing-library/jest-dom';

const mockSetSelectedData = jest.fn();
const mockSetUserAccount = jest.fn();
const mockSendData = jest.fn();

jest.mock('@src/services/AuthService/AuthService', () => ({
  sendData: (): void => mockSendData(),
}));

const initialProps: ModalType = {
  active: true,
  userId: '1',
  city: '',
  postalCode: '',
  country: '',
  streetName: '',
  setActive: jest.fn(),
  setSelectedData: mockSetSelectedData,
  setUserAccount: mockSetUserAccount,
  userAccount: {
    email: '',
    firstName: '',
    id: '',
    lastName: '',
    billingAddressIds: [''],
    shippingAddressIds: [''],
    defaultBillingAddressId: '',
    defaultShippingAddressId: '',
    addresses: [],
    dateOfBirth: '',
  },
  selectedData: {
    city: '',
    postalCode: '',
    country: '',
    streetName: '',
    addressId: '1',
  },
};

describe('Modal component', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Modal
        active={initialProps.active}
        userId={initialProps.userId}
        city={initialProps.city}
        postalCode={initialProps.postalCode}
        country={initialProps.country}
        streetName={initialProps.streetName}
        setActive={initialProps.setActive}
        setSelectedData={initialProps.setSelectedData}
        setUserAccount={initialProps.setUserAccount}
        userAccount={initialProps.userAccount}
        selectedData={initialProps.selectedData}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should handle input change', async () => {
    render(
      <Modal
        active={initialProps.active}
        userId={initialProps.userId}
        city={initialProps.city}
        postalCode={initialProps.postalCode}
        country={initialProps.country}
        streetName={initialProps.streetName}
        setActive={initialProps.setActive}
        setSelectedData={initialProps.setSelectedData}
        setUserAccount={initialProps.setUserAccount}
        userAccount={initialProps.userAccount}
        selectedData={initialProps.selectedData}
      />,
    );
    const cityInput = screen.getByLabelText(/City/i);
    fireEvent.change(cityInput, { target: { value: 'New York' } });
    await waitFor(() => {
      expect(mockSetSelectedData).toHaveBeenCalledWith(
        expect.objectContaining({
          city: 'New York',
        }),
      );
    });
  });

  it('should handle country select change', async () => {
    render(
      <Modal
        active={initialProps.active}
        userId={initialProps.userId}
        city={initialProps.city}
        postalCode={initialProps.postalCode}
        country={initialProps.country}
        streetName={initialProps.streetName}
        setActive={initialProps.setActive}
        setSelectedData={initialProps.setSelectedData}
        setUserAccount={initialProps.setUserAccount}
        userAccount={initialProps.userAccount}
        selectedData={initialProps.selectedData}
      />,
    );
    const countrySelect = screen.getByTestId('country');
    fireEvent.change(countrySelect, { target: { value: 'US' } });
    await waitFor(() => {
      expect(mockSetSelectedData).toHaveBeenCalledWith(
        expect.objectContaining({
          country: 'US',
        }),
      );
    });
  });
});
