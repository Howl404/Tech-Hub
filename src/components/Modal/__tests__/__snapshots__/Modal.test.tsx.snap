// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render correctly 1`] = `
<div>
  <div
    class="modal active__modal"
  >
    <form
      class="modal__content active__modal"
    >
      <div
        class="form-input"
      >
        <label
          for="country"
        >
          Country
          <select
            data-testid="country"
            id="country"
            name="country"
          >
            <option
              label="Select a country ... "
              value=""
            >
              Select a country ...
            </option>
            <option
              label="United States"
              value="US"
            >
              United States
            </option>
            <option
              label="Russia"
              value="RU"
            >
              Russia
            </option>
            <option
              label="United Kingdom"
              value="GB"
            >
              United Kingdom
            </option>
            <option
              label="Germany"
              value="DE"
            >
              Germany
            </option>
            <option
              label="France"
              value="FR"
            >
              France
            </option>
          </select>
        </label>
      </div>
      <div
        class="form-input"
      >
        <label
          for="city"
        >
          City
          <input
            id="city"
            max=""
            pattern="[A-Za-z]+"
            title="Must contain only letters"
            type="text"
            value=""
          />
          <span
            class="city-error"
          >
            City is not valid
          </span>
        </label>
      </div>
      <div
        class="form-input"
      >
        <label
          for="postalCode"
        >
          Postal Code
          <input
            id="postalCode"
            max=""
            pattern=".*"
            title="Must be a valid postal code of a selected country"
            type="text"
            value=""
          />
          <span
            class="postalCode-error"
          >
            Invalid Postal Code
          </span>
        </label>
      </div>
      <div
        class="form-input"
      >
        <label
          for="streetName"
        >
          Street name
          <input
            id="streetName"
            max=""
            pattern=".*"
            title="Must contain more than 1 character"
            type="text"
            value=""
          />
          <span
            class="streetName-error"
          >
            Less than 1 character
          </span>
        </label>
      </div>
      <button
        class="btn__save"
        disabled=""
        type="submit"
      >
        save 
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM272 80v80H144V80h128zm122 352H54a6 6 0 0 1-6-6V86a6 6 0 0 1 6-6h42v104c0 13.255 10.745 24 24 24h176c13.255 0 24-10.745 24-24V83.882l78.243 78.243a6 6 0 0 1 1.757 4.243V426a6 6 0 0 1-6 6zM224 232c-48.523 0-88 39.477-88 88s39.477 88 88 88 88-39.477 88-88-39.477-88-88-88zm0 128c-22.056 0-40-17.944-40-40s17.944-40 40-40 40 17.944 40 40-17.944 40-40 40z"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
